File: app/src/main/java/com/soumya/wwdablu/zomatobuddy/viewadapter/ReviewAdapter.java
package com.soumya.wwdablu.zomatobuddy.viewadapter;

import android.databinding.DataBindingUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.soumya.wwdablu.zomatobuddy.BuildConfig;
import com.soumya.wwdablu.zomatobuddy.R;
import com.soumya.wwdablu.zomatobuddy.databinding.CardReviewBinding;
import com.soumya.wwdablu.zomatobuddy.model.ReviewModel;
import com.soumya.wwdablu.zomatobuddy.model.reviews.ReviewResponse;
import com.soumya.wwdablu.zomatobuddy.model.reviews.Review;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;

public class ReviewAdapter extends RecyclerView.Adapter<ReviewAdapter.ReviewViewHolder> {

    private ReviewModel reviewModel;
    private DisposableObserver<ReviewResponse> disposableObserver;
    private ReviewResponse reviewReviewResponse;

    public ReviewAdapter(final String resId) {

        reviewModel = new ReviewModel(BuildConfig.ZOMATO_BASE_URL);

        disposableObserver = reviewModel.getReviews(resId)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeWith(new DisposableObserver<ReviewResponse>() {
                @Override
                public void onNext(ReviewResponse reviewResponse) {
                    reviewReviewResponse = reviewResponse;
                    notifyDataSetChanged();
                }

                @Override
                public void onError(Throwable e) {

                }

                @Override
                public void onComplete() {

                }
            });
    }

    @Override
    public ReviewViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {

        return new ReviewViewHolder(DataBindingUtil.inflate(
                LayoutInflater.from(parent.getContext()), R.layout.card_review,
                parent, false));
    }

    @Override
    public void onBindViewHolder(ReviewViewHolder holder, int position) {
        holder.onBind(position);
    }

    @Override
    public int getItemCount() {
        return null == reviewReviewResponse ? 0 : reviewReviewResponse.getUserReviews().size();
    }

    public void clean() {

        if(null != disposableObserver && !disposableObserver.isDisposed()) {
            disposableObserver.dispose();
        }
    }

    class ReviewViewHolder extends RecyclerView.ViewHolder {

        private CardReviewBinding binder;

        ReviewViewHolder(CardReviewBinding binder) {
            super(binder.getRoot());
            this.binder = binder;
        }

        void onBind(int position) {

            Review review = reviewReviewResponse.getUserReviews().get(position).getReview();

            binder.rbRating.setRating(((float) review.getRating()));
            binder.tvRatingText.setText(review.getRatingText());
            binder.tvReviewerName.setText(review.getUser().getName());
            binder.tvReviewDate.setText(review.getReviewTimeFriendly());
            binder.tvReviewContent.setText(review.getReviewText());
        }
    }
}

File: app/src/main/java/com/soumya/wwdablu/hungry/adapter/ReviewAdapter.kt
package com.soumya.wwdablu.hungry.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.soumya.wwdablu.hungry.R
import com.soumya.wwdablu.hungry.databinding.ItemReviewInfoBinding
import com.soumya.wwdablu.hungry.network.model.reviews.Review
import com.soumya.wwdablu.hungry.network.model.reviews.ReviewModel

class ReviewAdapter(reviewModel: ReviewModel) : BaseAdapter<ReviewAdapter.ReviewViewHolder>() {

    private val mReviewModel: ReviewModel = reviewModel

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ReviewViewHolder {

        return ReviewViewHolder(ItemReviewInfoBinding.inflate(LayoutInflater.from(parent.context),
                parent, false))
    }

    override fun onBindViewHolder(holder: ReviewViewHolder, position: Int) {
        holder.bind(mReviewModel.reviewList[position].review)
    }

    override fun getItemCount(): Int {
        return mReviewModel.reviewList.size
    }

    inner class ReviewViewHolder(viewBinding: ItemReviewInfoBinding) : RecyclerView.ViewHolder(viewBinding.root) {

        private val mViewBinding: ItemReviewInfoBinding = viewBinding

        fun bind(review: Review) {
            mViewBinding.review = review
            loadImageByUrl(mViewBinding.ivUserImage, review.user.profileImage, R.drawable.default_card_bg)
        }
    }
}